File: a982(maze_problem_1)/C++/maze_problem_1.cpp
#include <iostream>
#include <queue>


struct Path {
  Path(int x, int y, int step)
    :x(x)
    ,y(y)
    ,step(step)
  {}

  int x, y, step;
};


inline bool IsPath(const char block) {
  return (block == '.' ? true : false);
}


int main(int argc, char const *argv[]) {
  int n = 0;
  while (std::cin >> n) {
    bool maze[n][n];
    char block = ' ';
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < n; ++j) {
        std::cin >> block;
        maze[i][j] = IsPath(block);
      }
    }
    // path check in thr order of up -> right -> down -> left
    int directions[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    // this is for solution detecting
    bool exist_path = false;
    // data structure to do Breath-first search
    std::queue<Path> q;
    // From maze[1][1] to maze[n - 2][n - 2].
    // Run out of range won't happen since there's always the '#' boundary.
    // The maze path will be flip to '#'(false) after passing.
    q.emplace(1, 1, 1);
    while (!q.empty()) {
      const auto cur_path = q.front();
      q.pop();
      // If we reach the destination, search is over.
      if (cur_path.x == n - 2 && cur_path.y == n - 2) {
        std::cout << cur_path.step << '\n';
        exist_path = true;
        break;
      }

      // try 4 directions
      for (int i = 0; i < 4; ++i) {
        int cur_x = cur_path.x + directions[i][0],
            cur_y = cur_path.y + directions[i][1];
        // If there's a path, reach it!
        if (maze[cur_x][cur_y]) {
          q.emplace(cur_x, cur_y, cur_path.step + 1);
          // mark out
          maze[cur_x][cur_xy] = false;
        }
      }
    }
    if (!exist_path) {
      std::cout << "No solution!" << '\n';
    }
  }
  return 0;
}

File: a982(maze_problem_1)/Python/maze_problem_1.py
from collections import deque
from sys import stdin, stdout
from typing import List


class Path:
    def __init__(self, x: int, y: int, step: int) -> None:
        self.x: int = x
        self.y: int = y
        self.step: int = step


if __name__ == '__main__':
    while True:
        line: str = stdin.readline()
        if not line:
            break

        n: int = int(line)
        maze: List[List[bool]] = [([False] * n) for _ in range(n)]
        for i in range(n):
            blocks: str = stdin.readline()
            for j, block in enumerate(blocks):
                if block == '.':
                    maze[i][j] = True

        directions: List[List[int]] = [[-1, 0], [0, 1], [1, 0], [0, -1]]
        exist_path: bool = False

        q = deque()
        q.append(Path(1, 1, 1))
        while len(q):
            cur_path: Path = q.popleft()
            if cur_path.x == n - 2 and cur_path.y == n - 2:
                stdout.write(f'{cur_path.step}\n')
                exist_path = True
                break

            for x, y in directions:
                cur_x: int = cur_path.x + x
                cur_y: int = cur_path.y + y

                if maze[cur_x][cur_y]:
                    q.append(Path(cur_x, cur_y, cur_path.step + 1))
                    maze[cur_x][cur_y] = False

        if not exist_path:
                stdout.write('No solution!\n')
